File: sdk/typescript/src/result.ts
/**
 * Strongly typed Result type for error handling.
 *
 * Uses discriminated unions for exhaustive pattern matching.
 */

import type { SdkError } from "./error";

/**
 * Success result.
 */
export interface Ok<T> {
  readonly _tag: "Ok";
  readonly value: T;
}

/**
 * Error result.
 */
export interface Err<E> {
  readonly _tag: "Err";
  readonly error: E;
}

/**
 * Result type - either Ok or Err.
 */
export type Result<T, E = SdkError> = Ok<T> | Err<E>;

/**
 * Creates an Ok result.
 */
export function ok<T>(value: T): Ok<T> {
  return { _tag: "Ok", value };
}

/**
 * Creates an Err result.
 */
export function err<E>(error: E): Err<E> {
  return { _tag: "Err", error };
}

/**
 * Type guard for Ok results.
 */
export function isOk<T, E>(result: Result<T, E>): result is Ok<T> {
  return result._tag === "Ok";
}

/**
 * Type guard for Err results.
 */
export function isErr<T, E>(result: Result<T, E>): result is Err<E> {
  return result._tag === "Err";
}

/**
 * Maps the Ok value.
 */
export function map<T, U, E>(
  result: Result<T, E>,
  fn: (value: T) => U
): Result<U, E> {
  return isOk(result) ? ok(fn(result.value)) : result;
}

/**
 * Maps the Err value.
 */
export function mapErr<T, E, F>(
  result: Result<T, E>,
  fn: (error: E) => F
): Result<T, F> {
  return isErr(result) ? err(fn(result.error)) : result;
}

/**
 * Flat maps the Ok value.
 */
export function flatMap<T, U, E>(
  result: Result<T, E>,
  fn: (value: T) => Result<U, E>
): Result<U, E> {
  return isOk(result) ? fn(result.value) : result;
}

/**
 * Unwraps the Ok value or throws.
 */
export function unwrap<T, E>(result: Result<T, E>): T {
  if (isOk(result)) {
    return result.value;
  }
  throw new Error(`Unwrap called on Err: ${JSON.stringify(result.error)}`);
}

/**
 * Unwraps the Ok value or returns a default.
 */
export function unwrapOr<T, E>(result: Result<T, E>, defaultValue: T): T {
  return isOk(result) ? result.value : defaultValue;
}

/**
 * Unwraps the Ok value or computes a default.
 */
export function unwrapOrElse<T, E>(
  result: Result<T, E>,
  fn: (error: E) => T
): T {
  return isOk(result) ? result.value : fn(result.error);
}

/**
 * Pattern matches on a Result.
 */
export function match<T, E, U>(
  result: Result<T, E>,
  handlers: {
    ok: (value: T) => U;
    err: (error: E) => U;
  }
): U {
  return isOk(result) ? handlers.ok(result.value) : handlers.err(result.error);
}

/**
 * Converts a Promise to a Result.
 */
export async function fromPromise<T, E = SdkError>(
  promise: Promise<T>,
  mapError: (error: unknown) => E
): Promise<Result<T, E>> {
  try {
    const value = await promise;
    return ok(value);
  } catch (error) {
    return err(mapError(error));
  }
}

/**
 * Converts a Result to a Promise.
 */
export function toPromise<T, E>(result: Result<T, E>): Promise<T> {
  return isOk(result) ? Promise.resolve(result.value) : Promise.reject(result.error);
}

/**
 * Combines multiple Results into a single Result.
 */
export function all<T extends readonly unknown[], E>(
  results: { [K in keyof T]: Result<T[K], E> }
): Result<T, E> {
  const values: unknown[] = [];
  for (const result of results) {
    if (isErr(result)) {
      return result;
    }
    values.push(result.value);
  }
  return ok(values as unknown as T);
}

/**
 * Async Result type alias.
 */
export type AsyncResult<T, E = SdkError> = Promise<Result<T, E>>;

File: sdk/go/result.go
package sdk

// Result represents either a success value or an error.
// Uses Go generics for type safety.
type Result[T any] struct {
	value T
	err   error
	ok    bool
}

// Ok creates a successful Result.
func Ok[T any](value T) Result[T] {
	return Result[T]{value: value, ok: true}
}

// Err creates an error Result.
func Err[T any](err error) Result[T] {
	return Result[T]{err: err, ok: false}
}

// IsOk returns true if the Result is successful.
func (r Result[T]) IsOk() bool {
	return r.ok
}

// IsErr returns true if the Result is an error.
func (r Result[T]) IsErr() bool {
	return !r.ok
}

// Unwrap returns the value or panics if error.
func (r Result[T]) Unwrap() T {
	if !r.ok {
		panic("Unwrap called on Err result")
	}
	return r.value
}

// UnwrapOr returns the value or a default.
func (r Result[T]) UnwrapOr(defaultValue T) T {
	if r.ok {
		return r.value
	}
	return defaultValue
}

// UnwrapOrElse returns the value or computes a default.
func (r Result[T]) UnwrapOrElse(fn func(error) T) T {
	if r.ok {
		return r.value
	}
	return fn(r.err)
}

// Error returns the error or nil.
func (r Result[T]) Error() error {
	if r.ok {
		return nil
	}
	return r.err
}

// Value returns the value and ok status.
func (r Result[T]) Value() (T, bool) {
	return r.value, r.ok
}

// Match pattern matches on the Result.
func (r Result[T]) Match(onOk func(T), onErr func(error)) {
	if r.ok {
		onOk(r.value)
	} else {
		onErr(r.err)
	}
}

// Map transforms the success value.
func Map[T, U any](r Result[T], fn func(T) U) Result[U] {
	if r.ok {
		return Ok(fn(r.value))
	}
	return Err[U](r.err)
}

// MapErr transforms the error.
func MapErr[T any](r Result[T], fn func(error) error) Result[T] {
	if r.ok {
		return r
	}
	return Err[T](fn(r.err))
}

// FlatMap chains Result computations.
func FlatMap[T, U any](r Result[T], fn func(T) Result[U]) Result[U] {
	if r.ok {
		return fn(r.value)
	}
	return Err[U](r.err)
}

// All combines multiple Results into one.
func All[T any](results ...Result[T]) Result[[]T] {
	values := make([]T, 0, len(results))
	for _, r := range results {
		if !r.ok {
			return Err[[]T](r.err)
		}
		values = append(values, r.value)
	}
	return Ok(values)
}

// Collect collects Results from a slice.
func Collect[T any](items []T, fn func(T) Result[T]) Result[[]T] {
	results := make([]T, 0, len(items))
	for _, item := range items {
		r := fn(item)
		if !r.ok {
			return Err[[]T](r.err)
		}
		results = append(results, r.value)
	}
	return Ok(results)
}

// Try wraps a function that may panic into a Result.
func Try[T any](fn func() T) (result Result[T]) {
	defer func() {
		if r := recover(); r != nil {
			if err, ok := r.(error); ok {
				result = Err[T](err)
			} else {
				result = Err[T](NewError(ErrInternalError, "panic occurred"))
			}
		}
	}()
	return Ok(fn())
}

// FromError creates a Result from a value and error pair (Go idiom).
func FromError[T any](value T, err error) Result[T] {
	if err != nil {
		return Err[T](err)
	}
	return Ok(value)
}
